Pipe a mutable state parameter through CSS property parsing

weinig opened the pull request:
#### b7f46f82ce84b81fcb59a8f44ad3bdf22bb5a052
<pre>
Pipe a mutable state parameter through CSS property parsing
<a href="https://bugs.webkit.org/show_bug.cgi?id=290471">https://bugs.webkit.org/show_bug.cgi?id=290471</a>

Reviewed by Darin Adler.

This change replaces the CSSParserContext parameter passed through
CSS property parsing with a new mutable CSS::PropertyParserState
parameter. The original intent was to support the implementation
of CSS random(), which requires each random() function to have a
unique random index value for the currently parsed property, but
while implementing the change, it became clear the new parameter
could also be used to simplify primitive parsing by centralizing
the checks for quirky parsing and based on a current property field.

The following fields were removed from the `codegen-properties` dictionary:
    - parser-function-requires-additional-parameters
    - parser-function-requires-current-property
    - parser-function-requires-current-shorthand
    - parser-function-requires-value-pool

The following fields were added to the `codegen-properties` dictionary:
    - accepts-quirky-angle
    - accepts-quirky-color
    - accepts-quirky-length

This enables the generation of the following properties:
    - border-bottom-color
    - border-top-color
    - border-left-color
    - border-right-color
    - border-bottom-width
    - border-top-width
    - border-left-width
    - border-right-width
    - bottom
    - top
    - left
    - right
    - margin-bottom
    - margin-top
    - margin-left
    - margin-right

Additional changes were made to simplify and unify parsing, including:
    - Making all uses of the CSSParser class, which holds no real state, use
      static functions. A future pass should remove CSSParser entirely as the
      uses all call through to some other parser.
    - Moving all the color parsing functions to CSSPropertyParserConsumer+Color.h
      and renaming the one that doesn&apos;t take a context to indicate its &quot;deprecated
      do not use&quot; status.
    - Making the options to CSSPrimitiveValueResolver and MetaConsumer optional,
      and only actually needed when something non-standard is needed. This reduces
      a ton of code and makes them much easier to get right.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/animation/TimelineRange.cpp:
* Source/WebCore/animation/ViewTimeline.cpp:
* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
* Source/WebCore/css/CSSCounterStyleRule.cpp:
* Source/WebCore/css/CSSGroupingRule.cpp:
* Source/WebCore/css/CSSKeyframeRule.cpp:
* Source/WebCore/css/CSSKeyframesRule.cpp:
* Source/WebCore/css/CSSPageRule.cpp:
* Source/WebCore/css/CSSPendingSubstitutionValue.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/CSSStyleRule.cpp:
* Source/WebCore/css/CSSStyleSheet.cpp:
* Source/WebCore/css/CSSValuePool.cpp:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
* Source/WebCore/css/DOMCSSNamespace.cpp:
* Source/WebCore/css/FontFace.cpp:
* Source/WebCore/css/MutableStyleProperties.cpp:
* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/StyleSheetContents.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.h:
* Source/WebCore/css/calc/CSSCalcValue.cpp:
* Source/WebCore/css/calc/CSSCalcValue.h:
* Source/WebCore/css/parser/CSSParser.cpp:
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
* Source/WebCore/css/parser/CSSParserFastPaths.h:
* Source/WebCore/css/parser/CSSParserIdioms.cpp: Removed.
* Source/WebCore/css/parser/CSSParserIdioms.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSParserMode.h:
* Source/WebCore/css/parser/CSSParserTokenRangeGuard.h: Added.
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AngleDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AnglePercentageDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AppleVisualEffect.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AppleVisualEffect.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Attr.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Attr.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Background.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Background.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Box.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Box.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CSSPrimitiveValueResolver.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorAdjust.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorAdjust.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorInlines.h: Copied from Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorInterpolationMethod.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorInterpolationMethod.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Content.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Content.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CounterStyles.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CounterStyles.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Display.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Display.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Easing.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Easing.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+FlexDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+FrequencyDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Inline.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Inline.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Inset.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+IntegerDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthPercentageDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Lists.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Lists.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Masking.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Masking.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaConsumer.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaConsumerDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaResolver.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Motion.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Motion.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+NumberDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percentage.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percentage.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PercentageDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PositionTry.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PositionTry.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Primitives.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ratio.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ratio.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ResolutionDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+SVG.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+SVG.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ScrollSnap.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ScrollSnap.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Scrollbars.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Scrollbars.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Shapes.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Shapes.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Symbol.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+SymbolDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Syntax.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Syntax.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TimeDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transitions.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UI.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UI.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnicodeRange.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ViewTransition.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ViewTransition.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+WillChange.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+WillChange.h:
* Source/WebCore/css/parser/CSSPropertyParserOptions.h:
* Source/WebCore/css/parser/CSSPropertyParserState.h: Added.
* Source/WebCore/css/parser/CSSSupportsParser.cpp:
* Source/WebCore/css/query/GenericMediaQueryParser.cpp:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleProperties+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
* Source/WebCore/css/typedom/CSSUnitValue.cpp:
* Source/WebCore/html/ColorInputType.cpp:
* Source/WebCore/html/HTMLElement.cpp:
* Source/WebCore/html/HTMLMetaElement.cpp:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
* Source/WebCore/html/canvas/CanvasStyle.cpp:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
* Source/WebCore/page/IntersectionObserver.cpp:
* Source/WebCore/style/RuleSet.cpp:
* Source/WebCore/style/StyleResolver.cpp:
* Source/WebCore/svg/SVGLengthValue.h:
* Source/WebCore/svg/properties/SVGAnimationAdditiveValueFunctionImpl.cpp:
* Source/WebCore/svg/properties/SVGPropertyTraits.h:
* Source/WebCore/testing/Internals.cpp:
* Source/WebKit/UIProcess/API/wpe/WebKitColor.cpp:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
* Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:
* Tools/TestWebKitAPI/Tests/WebCore/CSSParserFastPaths.cpp:

Canonical link: <a href="https://commits.webkit.org/293017@main">https://commits.webkit.org/293017@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/425eda0301baec0c49cdd1ed9ab7587e64d73d43

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97722 "33 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17347 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7564 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102829 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48251 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99767 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17641 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25800 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/74431 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/31621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100725 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/13367 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/88338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/54780 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/97209 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13137 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6254 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47694 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/83139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6331 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104830 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24802 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83477 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25174 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84498 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82909 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/20882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27484 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5164 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18409 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24763 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29932 "Failed to checkout and rebase branch from PR 43066") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24585 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27899 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26159 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSKeyframesRule.cpp:
```diff
@@ -127,8 +128,7 @@ void CSSKeyframesRule::appendRule(const String& ruleText)
 
-    CSSParser parser(parserContext());
-    RefPtr<StyleRuleKeyframe> keyframe = parser.parseKeyframeRule(ruleText);
+    RefPtr<StyleRuleKeyframe> keyframe = CSSParser::parseKeyframeRule(ruleText, parserContext());
```
```suggestion
    auto keyframe = CSSParser::parseKeyframeRule(ruleText, parserContext());
```

darinadler commented on Source/WebCore/css/DOMCSSNamespace.cpp:
```diff
@@ -76,8 +76,8 @@ bool DOMCSSNamespace::supports(Document& document, const String& conditionText)
-    CSSParser parser(context);
-    return parser.parseSupportsCondition(conditionText);
+
+    return CSSParser::parseSupportsCondition(conditionText, context);
```
```suggestion
    return CSSParser::parseSupportsCondition(conditionText, context);
```

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.h:
```diff
@@ -47,13 +50,13 @@ enum TrackListType : uint8_t { GridTemplate, GridTemplateNoRepeat, GridAuto };
 
 bool isGridBreadthIdent(CSSValueID);
 bool parseGridTemplateAreasRow(StringView gridRowNames, NamedGridAreaMap&, const size_t rowCount, size_t& columnCount);
```
Spotted this in passing:
```suggestion
bool parseGridTemplateAreasRow(StringView gridRowNames, NamedGridAreaMap&, size_t rowCount, size_t& columnCount);
```

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.h:
```diff
@@ -47,13 +50,13 @@ enum TrackListType : uint8_t { GridTemplate, GridTemplateNoRepeat, GridAuto };
+RefPtr<CSSGridLineNamesValue> consumeGridLineNames(CSSParserTokenRange&, CSS::PropertyParserState&, AllowEmpty = AllowEmpty::No);
+RefPtr<CSSValue> consumeGridLine(CSSParserTokenRange&, CSS::PropertyParserState&);
+RefPtr<CSSValue> consumeGridTrackSize(CSSParserTokenRange&, CSS::PropertyParserState&);
+RefPtr<CSSValue> consumeGridTrackList(CSSParserTokenRange&, CSS::PropertyParserState&, TrackListType = TrackListType::GridAuto);
```
Is it better to have a track list type default? The slight risk would be that we’d forget to pass the type through. The benefit is concision if it’s common to call this for `GridAuto`.

weinig replied:
It was the only remaining place where `parser-function-requires-additional-parameters` was being used, so by adding the default parameter we could eliminate it. These will be generated soon enough, so I am going to leave it with the default for the moment.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserOptions.h:
```diff
@@ -26,20 +26,38 @@
     AnchorPolicy anchorPolicy                   { AnchorPolicy::Forbid };
+
+    // Generally, anchor-size() is forbidden, but in a few cases specified in CSS Anchor Positioning
+    // spec it can be allowed..
```
```suggestion
    // spec it can be allowed.
```

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserOptions.h:
```diff
@@ -26,20 +26,38 @@
+    std::optional<CSSParserMode> overrideParserMode { };
+
+    // Generally, anchor() is forbidden, but in a few cases specified in CSS Anchor Positioning
+    // spec it can be allowed..
```
```suggestion
    // spec it can be allowed.
```

darinadler commented on Source/WebCore/css/parser/CSSSupportsParser.cpp:
```diff
@@ -171,7 +172,12 @@ CSSSupportsParser::SupportsResult CSSSupportsParser::consumeSupportsSelectorFunc
+
+    auto state = CSS::PropertyParserState {
+        .context = m_parser.context(),
+    };
```
Seems great to go vertical and named for non-trivial cases, but for the trivial cases like this it seems sufficiently clear like this:
```suggestion
    auto state = CSS::PropertyParserState { m_parser.context() };
```

weinig replied:
Totally agree.

darinadler commented on Source/WebCore/html/ColorInputType.cpp:
```diff
@@ -34,6 +34,7 @@
 
 #include "AXObjectCache.h"
 #include "CSSPropertyParserConsumer+Color.h"
+#include "CSSPropertyParserConsumer+ColorInlines.h"
```
How do you feel about omitting the redundant include in a case like this?
```suggestion
#include "CSSPropertyParserConsumer+ColorInlines.h"
```

weinig replied:
Seems reasonable to me.

darinadler commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -834,7 +835,7 @@ std::optional<SRGBA<uint8_t>> HTMLElement::parseLegacyColorValue(StringView stri
         return std::nullopt;
 
-    if (auto namedColor = CSSParser::parseNamedColor(string))
+    if (auto namedColor = CSSParserFastPaths::parseNamedColor(string))
```
Seems a little bit of a naming mistake to have some _entry points_ into the parser be in a namespace `CSSParserFastPaths`. With a name like that I’d expect it to be an implementation detail of the parser, hidden behind a light abstraction.

weinig replied:
I agree. I am going to move all these color parsing functions in a follow up.

darinadler reviewed (approved):
Some day I hope many of these functions take `StringView` instead of `const String&`.

weinig committed:
Pipe a mutable state parameter through CSS property parsing
https://bugs.webkit.org/show_bug.cgi?id=290471

Reviewed by Darin Adler.

This change replaces the CSSParserContext parameter passed through
CSS property parsing with a new mutable CSS::PropertyParserState
parameter. The original intent was to support the implementation
of CSS random(), which requires each random() function to have a
unique random index value for the currently parsed property, but
while implementing the change, it became clear the new parameter
could also be used to simplify primitive parsing by centralizing
the checks for quirky parsing and based on a current property field.

The following fields were removed from the `codegen-properties` dictionary:
    - parser-function-requires-additional-parameters
    - parser-function-requires-current-property
    - parser-function-requires-current-shorthand
    - parser-function-requires-value-pool

The following fields were added to the `codegen-properties` dictionary:
    - accepts-quirky-angle
    - accepts-quirky-color
    - accepts-quirky-length

This enables the generation of the following properties:
    - border-bottom-color
    - border-top-color
    - border-left-color
    - border-right-color
    - border-bottom-width
    - border-top-width
    - border-left-width
    - border-right-width
    - bottom
    - top
    - left
    - right
    - margin-bottom
    - margin-top
    - margin-left
    - margin-right

Additional changes were made to simplify and unify parsing, including:
    - Making all uses of the CSSParser class, which holds no real state, use
      static functions. A future pass should remove CSSParser entirely as the
      uses all call through to some other parser.
    - Moving all the color parsing functions to CSSPropertyParserConsumer+Color.h
      and renaming the one that doesn't take a context to indicate its "deprecated
      do not use" status.
    - Making the options to CSSPrimitiveValueResolver and MetaConsumer optional,
      and only actually needed when something non-standard is needed. This reduces
      a ton of code and makes them much easier to get right.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/animation/TimelineRange.cpp:
* Source/WebCore/animation/ViewTimeline.cpp:
* Source/WebCore/contentextensions/ContentExtensionParser.cpp:
* Source/WebCore/css/CSSCounterStyleRule.cpp:
* Source/WebCore/css/CSSGroupingRule.cpp:
* Source/WebCore/css/CSSKeyframeRule.cpp:
* Source/WebCore/css/CSSKeyframesRule.cpp:
* Source/WebCore/css/CSSPageRule.cpp:
* Source/WebCore/css/CSSPendingSubstitutionValue.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/CSSStyleRule.cpp:
* Source/WebCore/css/CSSStyleSheet.cpp:
* Source/WebCore/css/CSSValuePool.cpp:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
* Source/WebCore/css/DOMCSSNamespace.cpp:
* Source/WebCore/css/FontFace.cpp:
* Source/WebCore/css/MutableStyleProperties.cpp:
* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/StyleSheetContents.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.h:
* Source/WebCore/css/calc/CSSCalcValue.cpp:
* Source/WebCore/css/calc/CSSCalcValue.h:
* Source/WebCore/css/parser/CSSParser.cpp:
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
* Source/WebCore/css/parser/CSSParserFastPaths.h:
* Source/WebCore/css/parser/CSSParserIdioms.cpp: Removed.
* Source/WebCore/css/parser/CSSParserIdioms.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSParserMode.h:
* Source/WebCore/css/parser/CSSParserTokenRangeGuard.h: Added.
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Align.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AngleDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AnglePercentageDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Animations.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AppleVisualEffect.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+AppleVisualEffect.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Attr.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Attr.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Background.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Background.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Box.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Box.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CSSPrimitiveValueResolver.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorAdjust.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorAdjust.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorInlines.h: Copied from Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorInterpolationMethod.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ColorInterpolationMethod.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Content.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Content.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CounterStyles.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CounterStyles.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Display.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Display.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Easing.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Easing.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+FlexDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+FrequencyDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Grid.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Inline.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Inline.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Inset.cpp: Removed.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+IntegerDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+LengthPercentageDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Lists.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Lists.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Masking.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Masking.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaConsumer.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaConsumerDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+MetaResolver.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Motion.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Motion.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+NumberDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percentage.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percentage.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PercentageDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PositionTry.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+PositionTry.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Primitives.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ratio.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ratio.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ResolutionDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+SVG.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+SVG.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ScrollSnap.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ScrollSnap.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Scrollbars.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Scrollbars.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Shapes.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Shapes.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Symbol.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+SymbolDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Syntax.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Syntax.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TimeDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Timeline.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transform.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transitions.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Transitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UI.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UI.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnicodeRange.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ViewTransition.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+ViewTransition.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+WillChange.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+WillChange.h:
* Source/WebCore/css/parser/CSSPropertyParserOptions.h:
* Source/WebCore/css/parser/CSSPropertyParserState.h: Added.
* Source/WebCore/css/parser/CSSSupportsParser.cpp:
* Source/WebCore/css/query/GenericMediaQueryParser.cpp:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleProperties+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
* Source/WebCore/css/typedom/CSSUnitValue.cpp:
* Source/WebCore/html/ColorInputType.cpp:
* Source/WebCore/html/HTMLElement.cpp:
* Source/WebCore/html/HTMLMetaElement.cpp:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
* Source/WebCore/html/canvas/CanvasStyle.cpp:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
* Source/WebCore/page/IntersectionObserver.cpp:
* Source/WebCore/style/RuleSet.cpp:
* Source/WebCore/style/StyleResolver.cpp:
* Source/WebCore/svg/SVGLengthValue.h:
* Source/WebCore/svg/properties/SVGAnimationAdditiveValueFunctionImpl.cpp:
* Source/WebCore/svg/properties/SVGPropertyTraits.h:
* Source/WebCore/testing/Internals.cpp:
* Source/WebKit/UIProcess/API/wpe/WebKitColor.cpp:
* Source/WebKit/WebProcess/InjectedBundle/API/c/WKBundlePage.cpp:
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
* Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:
* Tools/TestWebKitAPI/Tests/WebCore/CSSParserFastPaths.cpp:

Canonical link: https://commits.webkit.org/293017@main
